### ISSUE TITLE
query slow about near real time data

### ISSUE THREAD

**Issue by cxlRay** [apache/druid#9918]
when send batch query to Realtime processes tasks, the performance is too bad. TPS only have 80, response time is more than one second,
max response time can be 30 second and 99% response time can be 15 second.
what I confused is the data of Realtime processes are in mem, why response time of query is so long?

### Affected Version

v druid-0.16.1-incubating.

### Description

- Cluster size
coordinator and overlord: 2
historical: 7
middleManager: 7
broker: 5

- The testing tool is jmeter
- testing result
thread num: 500
TPS: 107.80
average response time: 4372ms
99% response time: 15559ms
max response time: 30150ms
min response time: 373
- the Flame chart of Realtime task code when test
![apm-flame](https://user-images.githubusercontent.com/11257788/82787626-67fc6300-9e99-11ea-92e0-08c28cf82ce2.png)

- the configuration of middleManager 
![middlManager](https://user-images.githubusercontent.com/11257788/82787935-10aac280-9e9a-11ea-9d92-bbfdf23bc400.png)

- the configuration of Realtime tasks
```
{
  "type": "kafka",
  "dataSchema": {
    "dataSource": "xxxx",
    "parser": {
      "type": "string",
      "parseSpec": {
        "format": "json",
        "timestampSpec": {
          "column": "timestamp",
          "format": "posix"
        },
        "dimensionsSpec": {
          "dimensions": ["tag1","tag2","tag3","tag4","tag5","tag6","tag7"],
          "dimensionExclusions": [
            "timestamp",
            "value"
          ]
        }
      }
    },
    "metricsSpec": [
      {
        "name": "value",
        "fieldName": "value",
        "type": "doubleSum"
      }
    ],
    "granularitySpec": {
      "type": "uniform",
      "segmentGranularity": "HOUR",
      "queryGranularity": "NONE",
      "rollup" : false
    }
  },
  "tuningConfig": {
    "type": "kafka",
    "intermediatePersistPeriod": "PT1H",
    "maxTotalRows": "245000000",
    "maxRowsPerSegment": 5000000
  },
  "ioConfig": {
    "topic": "xxxxx",
    "consumerProperties": {
      "bootstrap.servers": "xxxx:9092"
    },
    "taskCount": 16,
    "replicas": 1,
    "taskDuration": "PT1H"
  }
}
```

- about segments
![segment-1](https://user-images.githubusercontent.com/11257788/82791341-24592780-9ea0-11ea-9aff-bc206ce445ea.png)
![segments-2](https://user-images.githubusercontent.com/11257788/82791348-27ecae80-9ea0-11ea-8f7e-7e304bf4d40a.png)

**Comment by yuanlihan** [comment]
Hi @cxlRay,
You can try to minimise `intermediatePersistPeriod` which is `PT10M` by default and then the persisted data could apply the `"vectorize": "true"` option. Hope this helps to some extent.

**Comment by exherb** [comment]
some issue here.

**Comment by cxlRay** [comment]
@yuanlihan thanks for your answer, minimise intermediatePersistPeriod will create much more small file

**Comment by yuanlihan** [comment]
> @yuanlihan thanks for your answer, minimise intermediatePersistPeriod will create much more small file

@cxlRay that's true. But the temporary persist files will be cleaned up when the hourly tasks finished. And the persisted incremental files/indexes(with extra indexes to speed up query processing) will be more efficient than the in-memory `incrementalIndex`. As far as I know, when a query scan the latest in-memory `incrementalIndex`(like latest 10 minutes's data), Druid processes the in-memory fact table held by `incrementalIndex` row by row.
Actually, I had tried to minimise the value of `maxRowsInMemory` to reduce in-memory rows but this will also introduce some overheads caused by frequently persisting.

**Comment by exherb** [comment]
timeseries is very slow with filter. (10-20s)

**Comment by cxlRay** [comment]
@exherb say more about your question, if your cluster run on ssd, there will be better

**Comment by exherb** [comment]
> @exherb say more about your question, if your cluster run on ssd, there will be better

- historacal & middlemanager runs on ssd. 
- 5million rows per segement / kafka ingestion task
- use turnilo to query
- historical query/time is acceptable (around 100-500ms)
- peon query：timeseries(lookup) / topN query/time is acceptable
- peon query：topN(lookup)  / timeseries is very slow (timeseries query/time: 16s-40s)

**Comment by cxlRay** [comment]
just like **yuanlihan** say，minimise **intermediatePersistPeriod** or **maxRowsInMemory** , add query parmeter **vectorize**

**Comment by exherb** [comment]
> just like **yuanlihan** say，minimise **intermediatePersistPeriod** or **maxRowsInMemory** , add query parmeter **vectorize**

intermediatePersistPeriod: P10M
maxRowsInMemory: 1000000

`context: { vectorize: true }`

still slow.

**Comment by navis** [comment]
rows in no-rollup incremental index are stored as
```
ConcurrentSkipListMap<Long, ConcurrentLinkedDeque<IncrementalIndexRow>>
```
which seemed not easy to be fast, imho.

Anyway, can you try with coarser granularity, like 15 minutes or something?

**Comment by exherb** [comment]
> rows in no-rollup incremental index are stored as
> 
> ```
> ConcurrentSkipListMap<Long, ConcurrentLinkedDeque<IncrementalIndexRow>>
> ```
> 
> which seemed not easy to be fast, imho.
> 
> Anyway, can you try with coarser granularity, like 15 minutes or something?

we enabled rollup to 1minutes, segment granularity by hour. Are you suggestion change segment granularity to 15 minutes?

**Comment by cxlRay** [comment]
@exherb how do you know the slow part is **peon query**,the query process: **client** -> **broker** -> **indexing service(Peon)**, use monitor metric or analysis code?

**Comment by exherb** [comment]
> @exherb how do you know the slow part is **peon query**,the query process: **client** -> **broker** -> **indexing service(Peon)**, use monitor metric or analysis code?

by druid metrics: `query/segment/time` / `query/time` (datasource~=.+middlemanager.+)

**Comment by navis** [comment]
@exherb I mean `granularity` of timeseries query.

**Comment by louisliu318** [comment]
same issue.
